### ISSUE TITLE
Memory corruption with absl::variant on MSVC

### ISSUE THREAD

**Issue by cstanfill** [abseil/abseil-cpp#129]
I've been using absl::variant in a cross-platform cmake-based project and I've noticed random seeming stack corruption popping up around the code that's using absl::variant when I try to run it on MSVC. Valgrind and AddressSanitizer don't see anything amiss when I build on Linux. I have not tried MemorySanitizer yet.

I've reduced it to the following case:
https://github.com/cstanfill/absl-bug
Import into Visual Studio (I'm using Community - 15.7.2) as a CMake project and run the bug_test exe under x64-Release (RelWithDebugInfo). This is what I get: https://i.imgur.com/9qB4bvp.png

It seems the second disjunction in the variant must be something with a non-trivial destructor in order to trigger it.

**Comment by derekmauro** [comment]
I built your code using Bazel and MSVC in a Windows VM, but I couldn't reproduce the bug. Maybe there is some CMake difference that hopefully @zhangxy988 can take a look at soon.

**Comment by zhangxy988** [comment]
I was able to reproduce the bug. Confirmed the following:
 * the bug shows up in Release build, not Debug build
 * the bug is with `absl::variant`, not `std::variant` (it works fine with `/std:c++latest` which enables `std::variant`)
Still investigating the root cause, feels like a compiler bug to me.

**Comment by zhangxy988** [comment]
OK, so the root cause seems to be this function
```
template <class Self, std::size_t I>
VariantAccessResult<I, Self> AccessUnion(Self&& self, SizeT<I> /*i*/) {
  return reinterpret_cast<VariantAccessResult<I, Self>>(self);
}
```
The `reinterpret_cast` seems to be returning a *wrong* value. I already have a fix that gets the value in a different way (via recursive calls). I'll double check with my team and will put in the fix asap.

**Comment by cstanfill** [comment]
Great, glad to hear it. Until I can sync to head, `/std:c++latest` is a good workaround.

**Comment by zhangxy988** [comment]
Found a MSVC bug with the same root cause: https://developercommunity.visualstudio.com/content/problem/203664/wrong-assebly-code-is-generated-for-return-reinter.html
It seems adding `inline` to the function also fixes the bug. We will need to make a decision on which approach to use.

**Comment by zhangxy988** [comment]
This should be fixed by eb686c069f8a6aa87bd6f1ac344a330f361fe9ec.

**Comment by michygee23** [comment]
thanks guys
